src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/EntityMetadataProcessor.cs

```diff
@@ -0,0 +1,214 @@
+
+    private List<EntityMetadataAggregateHistory> _aggregatesToAdd = new();
+    private List<EntityMetadataHistory> _entriesToAdd = new();
+
+    public EntityMetadataProcessor(ProcessorContext context)
+    {
+        _context = context;
+    }
+
+    public void VisitUpsert(CoreModel.Substate substateData, ReferencedEntity referencedEntity, long stateVersion)
+    {
+        if (substateData is CoreModel.MetadataModuleEntrySubstate metadataEntry)
+        {
+            _changes
+                .GetOrAdd(new MetadataChangePointerLookup(referencedEntity.DatabaseId, stateVersion), _ => new MetadataChangePointer())
+                .Entries.Add(metadataEntry);
+        }
+    }
+
+    public void ProcessChanges()
+    {
+        foreach (var (lookup, change) in _changes.AsEnumerable())
+        {
+            EntityMetadataAggregateHistory aggregate;
+
+            if (!_mostRecentAggregates.TryGetValue(lookup.EntityId, out var previousAggregate) || previousAggregate.FromStateVersion != lookup.StateVersion)
+            {
+                aggregate = new EntityMetadataAggregateHistory
+                {
+                    Id = _context.Sequences.EntityMetadataAggregateHistorySequence++,
+                    FromStateVersion = lookup.StateVersion,
+                    EntityId = lookup.EntityId,
+                    MetadataIds = new List<long>(),
+                };
+
+                if (previousAggregate != null)
+                {
+                    aggregate.MetadataIds.AddRange(previousAggregate.MetadataIds);
+                }
+
+                _aggregatesToAdd.Add(aggregate);
+                _mostRecentAggregates[lookup.EntityId] = aggregate;
+            }
+            else
+            {
+                aggregate = previousAggregate;
+            }
+
+            foreach (var entry in change.Entries)
+            {
+                var entryLookup = new MetadataEntryDbLookup(lookup.EntityId, entry.Key.Name);
+                var entryHistory = new EntityMetadataHistory
+                {
+                    Id = _context.Sequences.EntityMetadataHistorySequence++,
+                    FromStateVersion = lookup.StateVersion,
+                    EntityId = lookup.EntityId,
+                    Key = entry.Key.Name,
+                    Value = entry.Value?.DataStruct.StructData.Hex.ConvertFromHex(),
+                    IsDeleted = entry.Value == null,
+                    IsLocked = entry.IsLocked,
+                };
+
+                _entriesToAdd.Add(entryHistory);
+
+                if (_mostRecentEntries.TryGetValue(entryLookup, out var previousEntry))
+                {
+                    var currentPosition = aggregate.MetadataIds.IndexOf(previousEntry.Id);
+
+                    if (currentPosition != -1)
+                    {
+                        aggregate.MetadataIds.RemoveAt(currentPosition);
+                    }
+                }
+
+                // TODO introduce entry.IsDeleted extension method
```

Comment:
do we have any task for that? can we maybe add it to this pullrequest? I'm affraid that we'll just leave that todo here forever, so either
1. create a task for that
2. code it in this pr
3. remove todo.

Reply:
I've literally moved as much existing code as possible. I don't think I'd like to change every todo right now.

Reply:
we had that conversation already :P we decided to either 
1. track these todo by jira task
2. leave that todo, if you're 100% sure that it'll be addressed in very short term. (not a big fan of that option though, it should never hit develop branch and wip tasks can be tracked other way)
3. remove it

Reply:
Removed it.